src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -157,15 +256,295 @@ private static TensorFlowTransform Create(IHostEnvironment env, ModelLoadContext
+
+            inputsForTraining[inputsForTraining.Length - 1] = args.LabeLColumn;
+
+            var inputSchema = input.Schema;
+            for (int i = 0; i < inputsForTraining.Length; i++)
+            {
+                if (!inputSchema.TryGetColumnIndex(inputsForTraining[i], out inputColIndices[i]))
+                    throw _host.Except($"Column {inputsForTraining[i]} doesn't exist");
+
+                var type = inputSchema.GetColumnType(inputColIndices[i]);
+                isInputVector[i] = type.IsVector;
+
+                var tfInput = new TFOutput(Graph[inputsForTraining[i]]);
+                tfInputTypes[i] = tfInput.OutputType;
+                tfInputShapes[i] = Graph.GetTensorShape(tfInput);
+                if (tfInputShapes[i].NumDimensions != -1)
+                {
+                    var newShape = new long[tfInputShapes[i].NumDimensions];
+                    newShape[0] = tfInputShapes[i][0] == -1 ? args.BatchSize : tfInputShapes[i][0];
+
+                    for (int j = 1; j < tfInputShapes[i].NumDimensions; j++)
+                        newShape[j] = tfInputShapes[i][j];
+                    tfInputShapes[i] = new TFShape(newShape);
+                }
+
+                var expectedType = TensorFlowUtils.Tf2MlNetType(tfInputTypes[i]);
+                if (type.ItemType != expectedType)
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", inputsForTraining[i], expectedType.ToString(), type.ToString());
+            }
+
+            var fetchList = new List<string>();
+            if (args.LossOperation != null)
+                fetchList.Add(args.LossOperation);
+            if (args.MetricOperation != null)
+                fetchList.Add(args.MetricOperation);
+
+            var hashedIndices = new HashSet<int>(inputColIndices);
+            for (int epoch = 0; epoch < args.Epoch; epoch++)
+            {
+                using (var cursor = input.GetRowCursor(a => hashedIndices.Contains(a)))
+                {
+                    var srcTensorGetters = GetTensorValueGetters(cursor, inputColIndices, isInputVector, tfInputTypes, tfInputShapes);
+
+                    float loss=0;
+                    float metric=0;
+                    bool isDataLeft = false;
+                    using (var ch = _host.Start("Training TensorFlow model..."))
+                    using (var pch = _host.StartProgressChannel("TensorFlow training progress..."))
+                    {
+                        pch.SetHeader(new ProgressHeader("Epoch"), (e) => e.SetProgress(0, epoch));
+                        if (args.LossOperation != null)
+                            pch.SetHeader(new ProgressHeader("Average Loss"), (e) => e.SetProgress(1, loss / ((cursor.Position + 1) / args.BatchSize)));
+                        if (args.MetricOperation != null)
+                            pch.SetHeader(new ProgressHeader("Average Metric"), (e) => e.SetProgress(2, metric / ((cursor.Position + 1) / args.BatchSize)));
```

Comment:
/// A common interface for progress reporting.
    /// It is expected that the progress channel interface is used from only one thread.
    ///
    /// Supported workflow:
    /// 1) Create the channel via <see cref="IProgressChannelProvider.StartProgressChannel"/>.
    /// 2) Call <see cref="SetHeader"/> as many times as desired (including 0).
    ///       Each call to <see cref="SetHeader"/> supersedes the previous one.
    /// 3) Report checkpoints (0 or more) by calling <see cref="Checkpoint"/>.
    /// 4) Repeat steps 2-3 as often as necessary.
    /// 5) Dispose the channel.

see 2).

I think you need to do something like this:

pch.SetHeader(new ProgressHeader(metrics.ToArray() new[] { "Epochs" }, entry => entry.SetProgress(0, epoch, _args.Epochs);

and after each train
if (metric.Count>0)
pch.Checkpoint(double[] avgMetrics);
where metrics and avgMetrics depend on  your arguments.